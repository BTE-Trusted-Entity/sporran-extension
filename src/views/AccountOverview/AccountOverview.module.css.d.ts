declare namespace AccountOverviewModuleCssNamespace {
  export interface IAccountOverviewModuleCss {
    button: string;
    container: string;
    heading: string;
    import: string;
    info: string;
  }
}

declare const AccountOverviewModuleCssModule: AccountOverviewModuleCssNamespace.IAccountOverviewModuleCss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: AccountOverviewModuleCssNamespace.IAccountOverviewModuleCss;
};

export = AccountOverviewModuleCssModule;
