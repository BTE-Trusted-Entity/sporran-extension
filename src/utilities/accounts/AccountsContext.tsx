import { createContext } from 'react';
import useSWR from 'swr';

import { AccountsMap } from './types';
import { getAccounts } from './getAccounts';

export interface AccountsContextType {
  data?: AccountsMap;
  error?: string;
}

export const AccountsContext = createContext<AccountsContextType>({});

export function AccountsProvider({
  children,
}: {
  children: JSX.Element;
}): JSX.Element {
  const value = useSWR('accounts', getAccounts);
  return (
    <AccountsContext.Provider value={value}>
      {children}
    </AccountsContext.Provider>
  );
}
