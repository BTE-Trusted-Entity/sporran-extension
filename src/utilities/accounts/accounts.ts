import { browser } from 'webextension-polyfill-ts';
import useSWR, { mutate, SWRResponse } from 'swr';
import { Identity } from '@kiltprotocol/core';

import { saveEncrypted } from '../storageEncryption/storageEncryption';

const storage = browser.storage.local;

const ACCOUNTS_KEY = 'accounts';
const CURRENT_ACCOUNT_KEY = 'currentAccount';

export interface Account {
  address: string;
  name: string;
}

export type AccountsMap = Record<string, Account>;

async function getAccounts(): Promise<AccountsMap> {
  return (await storage.get(ACCOUNTS_KEY))[ACCOUNTS_KEY] || {};
}

export function useAccounts(): SWRResponse<AccountsMap, unknown> {
  return useSWR(ACCOUNTS_KEY, getAccounts);
}

async function getCurrentAccount(): Promise<string | null> {
  const stored = await storage.get([ACCOUNTS_KEY, CURRENT_ACCOUNT_KEY]);
  const accounts = stored[ACCOUNTS_KEY] as AccountsMap;
  const current = stored[CURRENT_ACCOUNT_KEY];

  if (accounts[current]) {
    return current;
  }

  const firstAccount = Object.values(accounts)[0];
  if (!firstAccount) {
    return null;
  }

  await setCurrentAccount(firstAccount.address);

  return firstAccount.address;
}

export async function setCurrentAccount(address: string): Promise<void> {
  const oldAddress = await storage.get(CURRENT_ACCOUNT_KEY);
  if (address === oldAddress[CURRENT_ACCOUNT_KEY]) {
    return;
  }
  await storage.set({ [CURRENT_ACCOUNT_KEY]: address });
  await mutate(CURRENT_ACCOUNT_KEY);
}

export function useCurrentAccount(): SWRResponse<string | null, unknown> {
  return useSWR(CURRENT_ACCOUNT_KEY, getCurrentAccount);
}

export async function saveAccount(account: Account): Promise<void> {
  const accounts = await getAccounts();
  accounts[account.address] = account;
  await storage.set({ [ACCOUNTS_KEY]: accounts });
  await mutate(ACCOUNTS_KEY);
}

export async function createAccount(
  backupPhrase: string,
  password: string,
): Promise<Account> {
  const { address, seed } = Identity.buildFromMnemonic(backupPhrase);
  await saveEncrypted(address, password, seed);

  const name = 'My Sporran Account';
  const account = { name, address };
  await saveAccount(account);
  return account;
}
