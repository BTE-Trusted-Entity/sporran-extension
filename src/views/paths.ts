export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
