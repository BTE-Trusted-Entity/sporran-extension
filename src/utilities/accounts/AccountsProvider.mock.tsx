import { AccountsMap } from './accounts';
import { AccountsContext } from './AccountsContext';

export const accountsMock: AccountsMap = {
  '4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire': {
    name: 'My Sporran Account',
    address: '4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire',
    index: 1,
  },
  '4sm9oDiYFe22D7Ck2aBy5Y2gzxi2HhmGML98W9ZD2qmsqKCr': {
    name: 'My Second Account',
    address: '4sm9oDiYFe22D7Ck2aBy5Y2gzxi2HhmGML98W9ZD2qmsqKCr',
    index: 2,
  },
  '4oyRTDhHL22Chv9T89Vv2TanfUxFzBnPeMuq4EFL3gUiHbtL': {
    name: 'My Third Account which has a very long name and might not even fit on the screen',
    address: '4oyRTDhHL22Chv9T89Vv2TanfUxFzBnPeMuq4EFL3gUiHbtL',
    index: 3,
  },
};

export const moreAccountsMock: AccountsMap = {
  ...accountsMock,
  '4ruKeJZXBuqvgTvsTpbsG1RChkTsdz1TDMGgFP7SYykK78R8': {
    name: 'My Fourth Account',
    address: '4ruKeJZXBuqvgTvsTpbsG1RChkTsdz1TDMGgFP7SYykK78R8',
    index: 4,
  },
  '4o9GN2oCv8E9GVrLMoYRgVWBYAtjXfWieKECsZmwKXevdSiV': {
    name: 'My Fifth Account',
    address: '4o9GN2oCv8E9GVrLMoYRgVWBYAtjXfWieKECsZmwKXevdSiV',
    index: 5,
  },
  '4srs2Ag4NQJyr9uDszokjT4EdiHfDHiv2hzvuZtJL7KFCVWo': {
    name: 'My Sixth Account',
    address: '4srs2Ag4NQJyr9uDszokjT4EdiHfDHiv2hzvuZtJL7KFCVWo',
    index: 6,
  },
};

export function AccountsProviderMock({
  accounts = accountsMock,
  children,
}: {
  accounts?: AccountsMap;
  children: JSX.Element;
}): JSX.Element {
  return (
    <AccountsContext.Provider value={{ data: accounts }}>
      {children}
    </AccountsContext.Provider>
  );
}
