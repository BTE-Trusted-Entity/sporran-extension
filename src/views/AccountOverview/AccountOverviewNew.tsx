import { browser } from 'webextension-polyfill-ts';
import { useRouteMatch } from 'react-router-dom';

import { plural } from '../../utilities/plural/plural';
import { NEW, useAccounts } from '../../utilities/accounts/accounts';
import { AccountsCarousel } from '../../components/AccountsCarousel/AccountsCarousel';
import { Stats } from '../../components/Stats/Stats';

import styles from './AccountOverview.module.css';

interface Props {
  nextTartan: string;
}

export function AccountOverviewNew({ nextTartan }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;
  const { path } = useRouteMatch();

  const accounts = useAccounts().data;
  if (!accounts) {
    return null;
  }

  const accountsNumber = Object.values(accounts).length;

  return (
    <main className={styles.container}>
      <header>
        <h1 className={styles.heading}>{t('view_AccountOverview_title')}</h1>
        <p className={styles.info}>
          {plural(accountsNumber, {
            one: 'view_AccountOverview_subtitle_one',
            other: 'view_AccountOverview_subtitle_other',
          })}
        </p>
      </header>

      <AccountsCarousel path={path} account={NEW} nextTartan={nextTartan} />

      <Stats />
    </main>
  );
}
