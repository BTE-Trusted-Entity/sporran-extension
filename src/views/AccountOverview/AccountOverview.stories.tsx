import { Meta } from '@storybook/react';
import { MemoryRouter, Route } from 'react-router-dom';

import { accountsMock as accounts } from '../../testing/AccountsProviderMock';
import { NEW } from '../../utilities/accounts/accounts';
import { paths } from '../paths';

import { AccountOverview } from './AccountOverview';

export default {
  title: 'Views/AccountOverview',
  component: AccountOverview,
} as Meta;

const address = '/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire';
const account = accounts['4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire'];

export function Template(): JSX.Element {
  return (
    <MemoryRouter initialEntries={[address]}>
      <Route path={paths.account.overview}>
        <AccountOverview account={account} />
      </Route>
    </MemoryRouter>
  );
}

export function New(): JSX.Element {
  return (
    <MemoryRouter initialEntries={['/account/NEW/send']}>
      <Route path={paths.account.overview}>
        <AccountOverview account={NEW} />
      </Route>
    </MemoryRouter>
  );
}

export function CreateSuccess(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        '/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire/created',
        address,
      ]}
    >
      <Route path={paths.account.overview}>
        <AccountOverview account={account} />
      </Route>
    </MemoryRouter>
  );
}

export function ImportSuccess(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        '/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire/imported',
        address,
      ]}
    >
      <Route path={paths.account.overview}>
        <AccountOverview account={account} />
      </Route>
    </MemoryRouter>
  );
}

export function ResetSuccess(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        '/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire/reset',
        address,
      ]}
    >
      <Route path={paths.account.overview}>
        <AccountOverview account={account} />
      </Route>
    </MemoryRouter>
  );
}
