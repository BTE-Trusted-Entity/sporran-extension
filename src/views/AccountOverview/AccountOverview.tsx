import { useCallback, useState } from 'react';
import { browser } from 'webextension-polyfill-ts';
import { Link, useParams, useRouteMatch, Redirect } from 'react-router-dom';
import { minBy } from 'lodash-es';

import { AccountsCarousel } from '../../components/AccountsCarousel/AccountsCarousel';
import { Balance } from '../../components/Balance/Balance';
import { Stats } from '../../components/Stats/Stats';
import { AccountSuccessOverlay } from '../../components/AccountSuccessOverlay/AccountSuccessOverlay';
import { Account, isNew, useAccounts } from '../../utilities/accounts/accounts';
import { plural } from '../../utilities/plural/plural';
import { generatePath, paths } from '../paths';
import { AccountOverviewNew } from './AccountOverviewNew';

import styles from './AccountOverview.module.css';

interface Props {
  account: Account;
}

export function AccountOverview({ account }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;
  const { path } = useRouteMatch();
  const params = useParams() as { type?: 'created' | 'imported' | 'pwreset' };

  const [hasOpenOverlay, setHasOpenOverlay] = useState(Boolean(params.type));
  const [type] = useState(params.type);

  const handleSuccessOverlayButtonClick = useCallback(() => {
    setHasOpenOverlay(false);
  }, []);

  const accounts = useAccounts().data;
  if (!accounts) {
    return null;
  }

  const credentialsAccount = minBy(Object.values(accounts), 'index') as Account;
  const accountsNumber = Object.values(accounts).length;
  const { address } = account;

  if (params.type) {
    return <Redirect to={generatePath(paths.account.overview, { address })} />;
  }

  if (isNew(account)) {
    return <AccountOverviewNew />;
  }

  return (
    <main className={styles.container}>
      <header>
        <h1 className={styles.heading}>{t('view_AccountOverview_title')}</h1>
        <p className={styles.info}>
          {plural(accountsNumber, {
            one: 'view_AccountOverview_subtitle_one',
            other: 'view_AccountOverview_subtitle_other',
          })}
        </p>
      </header>

      <AccountsCarousel path={path} account={account} />

      <Balance address={address} breakdown />

      <p>
        <Link
          to={generatePath(paths.account.send.start, { address })}
          className={styles.button}
        >
          {t('view_AccountOverview_send')}
        </Link>

        <Link
          to={generatePath(paths.account.receive, { address })}
          className={styles.button}
        >
          {t('view_AccountOverview_receive')}
        </Link>
      </p>

      {account === credentialsAccount && (
        <Link
          to={generatePath(paths.account.credentials, { address })}
          className={styles.credentials}
        >
          {t('view_AccountOverview_credentials')}
        </Link>
      )}

      <Stats />
      {hasOpenOverlay && type && (
        <AccountSuccessOverlay
          successType={type}
          account={account}
          onSuccessOverlayButtonClick={handleSuccessOverlayButtonClick}
        />
      )}
    </main>
  );
}
