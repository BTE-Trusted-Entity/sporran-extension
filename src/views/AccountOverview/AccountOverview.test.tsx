import { MemoryRouter, Route } from 'react-router-dom';

import {
  accountsMock,
  mockBackgroundScript,
  render,
} from '../../testing/testing';
import { waitForNextTartan } from '../../utilities/accounts/getNextTartan.mock';
import { NEW } from '../../utilities/accounts/accounts';
import { paths } from '../paths';

import { AccountOverview } from './AccountOverview';

const account =
  accountsMock['4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire'];

describe('AccountOverview', () => {
  it('should render a normal account', async () => {
    mockBackgroundScript();

    const { container } = render(
      <MemoryRouter initialEntries={[`/account/${account.address}/`]}>
        <Route path={paths.account.overview}>
          <AccountOverview account={account} />,
        </Route>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the new account', async () => {
    mockBackgroundScript();

    const { container } = render(
      <MemoryRouter initialEntries={['/account/NEW/']}>
        <Route path={paths.account.overview}>
          <AccountOverview account={NEW} />,
        </Route>
      </MemoryRouter>,
    );
    await waitForNextTartan();
    expect(container).toMatchSnapshot();
  });
});
