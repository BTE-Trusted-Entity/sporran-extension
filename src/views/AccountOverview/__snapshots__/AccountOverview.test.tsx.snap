// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountOverview should render a normal account 1`] = `
<div>
  <main>
    <header>
      <h1>
        Your accounts
      </h1>
      <p>
        You have $number$ accounts
      </p>
    </header>
    <a
      aria-label="Create new account"
      href="/account/NEW"
      title="Create new account"
    >
      ←
    </a>
    <section>
      <h2
        style="display: inline;"
      >
        My Sporran Account
      </h2>
      <nav
        class="container"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="Account options"
          class="button"
          role="button"
          tabindex="0"
          title="Account options"
          type="button"
        >
          v
        </button>
      </nav>
    </section>
    <a
      aria-label="My Second Account"
      href="/account/4sm9oDiYFe22D7Ck2aBy5Y2gzxi2HhmGML98W9ZD2qmsqKCr"
      title="My Second Account"
    >
      →
    </a>
    <p>
      <span>
        Balance: 
        1.2340
         
        K
      </span>
    </p>
    <p>
      <a
        href="/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire/send"
      >
        Send
      </a>
    </p>
    <p>
      <a
        href="/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire/receive"
      >
        Receive
      </a>
    </p>
    <p>
      You have $number$ accounts with a total
       Balance: 
      3.7020
       
      K
    </p>
  </main>
  ,
</div>
`;

exports[`AccountOverview should render the new account 1`] = `
<div>
  <main>
    <header>
      <h1>
        Your accounts
      </h1>
      <p>
        You have $number$ accounts
      </p>
    </header>
    <a
      aria-label="My Third Account"
      href="/account/4oyRTDhHL22Chv9T89Vv2TanfUxFzBnPeMuq4EFL3gUiHbtL"
      title="My Third Account"
    >
      ←
    </a>
    <section>
      <a
        href="/account/create"
      >
        <h2>
          Create new account
        </h2>
      </a>
      <p>
        or
      </p>
      <a
        href="/account/import"
      >
        Import an account from pre-existing phrase
      </a>
    </section>
    <a
      aria-label="My Sporran Account"
      href="/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire"
      title="My Sporran Account"
    >
      →
    </a>
    <p>
      You have $number$ accounts with a total
       Balance: 
      3.7020
       
      K
    </p>
  </main>
  ,
</div>
`;
